File: ios/Classes/EzoicBannerPlatformView.swift
import Flutter
import UIKit
import EzoicAdsSDKBinary

class EzoicBannerViewFactory: NSObject, FlutterPlatformViewFactory {
  private let messenger: FlutterBinaryMessenger
  init(messenger: FlutterBinaryMessenger) { self.messenger = messenger }

  func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
    return EzoicBannerPlatformView(frame: frame, viewId: viewId, args: args, messenger: messenger)
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol { FlutterStandardMessageCodec.sharedInstance() }
}

class EzoicBannerPlatformView: NSObject, FlutterPlatformView, EzoicBannerViewDelegate {
  private let container = UIView()
  private let channel: FlutterMethodChannel
  private var banner: EzoicBannerView?

  init(frame: CGRect, viewId: Int64, args: Any?, messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: "com.ezoic/ezoic_banner_view_\(viewId)", binaryMessenger: messenger)
    super.init()
    let params = args as? [String: Any]
    let adUnitId = Int(params?["adUnitIdentifier"] as? String ?? "") ?? 0
    let sizes = (params?["size"] as? String ?? "").split(separator: ",").map { String($0) }
    let view = EzoicBannerView(adUnitIdentifier: adUnitId)
    view.delegate = self
    view.translatesAutoresizingMaskIntoConstraints = false
    container.addSubview(view)
    NSLayoutConstraint.activate([
      view.centerXAnchor.constraint(equalTo: container.centerXAnchor),
      view.centerYAnchor.constraint(equalTo: container.centerYAnchor),
    ])
    banner = view
    if sizes.isEmpty { view.loadAd() } else { view.loadAd(sizes: sizes) }
  }

  func view() -> UIView { container }

  func bannerViewDidLoad(_ bannerView: EzoicBannerView) { channel.invokeMethod("onLoad", arguments: nil) }
  func bannerView(_ bannerView: EzoicBannerView, didFailToLoadWithError error: EzoicError) {
    channel.invokeMethod("onError", arguments: ["message": error.localizedDescription, "code": error.code])
  }
  func bannerViewDidRecordImpression(_ bannerView: EzoicBannerView) { channel.invokeMethod("onImpression", arguments: nil) }
  func bannerViewDidRecordClick(_ bannerView: EzoicBannerView) { channel.invokeMethod("onClick", arguments: nil) }
  func bannerViewWillPresentScreen(_ bannerView: EzoicBannerView) { channel.invokeMethod("onOpen", arguments: nil) }
  func bannerViewDidDismissScreen(_ bannerView: EzoicBannerView) { channel.invokeMethod("onClose", arguments: nil) }
}

File: android/src/main/kotlin/com/ezoic/ezoic_flutter_sdk/EzoicBannerPlatformView.kt
package com.ezoic.ezoic_flutter_sdk

import android.content.Context
import android.view.View
import android.widget.FrameLayout
import com.ezoic.ads.sdk.adunits.EzoicBannerView
import com.ezoic.ads.sdk.adunits.EzoicBannerViewListener
import com.ezoic.ads.sdk.core.EzoicError
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class EzoicBannerViewFactory(private val messenger: BinaryMessenger) :
  PlatformViewFactory(StandardMessageCodec.INSTANCE) {
  override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
    @Suppress("UNCHECKED_CAST")
    val params = (args as? Map<String, Any?>) ?: emptyMap()
    return EzoicBannerPlatformView(context, viewId, params, messenger)
  }
}

class EzoicBannerPlatformView(
  context: Context,
  viewId: Int,
  params: Map<String, Any?>,
  messenger: BinaryMessenger,
) : PlatformView {

  private val container = FrameLayout(context)
  private val channel = MethodChannel(messenger, "com.ezoic/ezoic_banner_view_$viewId")

  init {
    val adUnitId = (params["adUnitIdentifier"] as? String)?.toIntOrNull() ?: 0
    val sizes = (params["size"] as? String ?: "").split(",").map { it.trim() }.filter { it.isNotEmpty() }
    val banner = EzoicBannerView(context, adUnitId)
    banner.listener = object : EzoicBannerViewListener {
      override fun onBannerLoaded(b: EzoicBannerView) { channel.invokeMethod("onLoad", null) }
      override fun onBannerLoadFailed(b: EzoicBannerView, error: EzoicError) {
        channel.invokeMethod("onError", mapOf("message" to error.message, "code" to error.code))
      }
      override fun onBannerImpression(b: EzoicBannerView) { channel.invokeMethod("onImpression", null) }
      override fun onBannerClicked(b: EzoicBannerView) { channel.invokeMethod("onClick", null) }
      override fun onBannerOpened(b: EzoicBannerView) { channel.invokeMethod("onOpen", null) }
      override fun onBannerClosed(b: EzoicBannerView) { channel.invokeMethod("onClose", null) }
    }
    container.addView(banner)
    if (sizes.isEmpty()) banner.loadAd() else banner.loadAd(sizes)
  }

  override fun getView(): View = container
  override fun dispose() {}
}
